src/postgresql/main/PostgreSqlMigrationStorage.cs

```diff
@@ -0,0 +1,115 @@
+                new {
+                    MigrationName = migrationName
+                });
+            return migrationInfo;
+        }
+
+        public async Task UpsertMigrationInfoAsync(IMigrationContext context, MigrationInfo info)
+        {
+            var db = GetDb(context);
+            long migrationInfoId = 0;
+            string insertQuery = $@"INSERT INTO {PostgreSqlConstants.MigrationInfoTableName}(
+                                        Name,
+                                        MigrationCompleted,
+                                        TotalMigrationTimeInMs,
+                                        CompletedAtUtc) 
+                                   VALUES (
+                                        @Name,
+                                        @MigrationCompleted,
+                                        @TotalMigrationTimeInMs,
+                                        @CompletedAtUtc)
+                                   ON CONFLICT (name)
+                                   DO UPDATE SET
+                                        Name = @Name,
+                                        MigrationCompleted = @MigrationCompleted,
+                                        TotalMigrationTimeInMs = @TotalMigrationTimeInMs,
+                                        CompletedAtUtc = @CompletedAtUtc
+                                   RETURNING id";
+
+            using (var reader = await db.ExecuteReaderAsync(insertQuery, new
+            {
+                Name = info.Name,
+                MigrationCompleted = info.MigrationCompleted,
+                TotalMigrationTimeInMs = info.TotalMigrationTimeInMs,
+                CompletedAtUtc = info.CompletedAtUtc
+            }))
+            {
+                while (reader.Read())
+                {
+                    migrationInfoId = reader.GetInt64(0);
+                }
+            }
+            foreach (var stepName in info.StepsCompleted)
+            {
+                string completedStepInsertQuery = $@"INSERT INTO {PostgreSqlConstants.CompletedStepsTableName}(
+                                        StepName,
+                                        MigrationInfoId)
+                                   VALUES (
+                                        @StepName,
+                                        @MigrationInfoId)
+                                   ON CONFLICT (StepName, MigrationInfoId)
+                                   DO NOTHING
+                                    ";
+
+                await db.ExecuteAsync(completedStepInsertQuery, new
+                {
+                    StepName = stepName,
+                    MigrationInfoId = migrationInfoId
+                });
+
+            }
+        }
+
+        public async Task<bool> HasMigrationBeenFullyCompletedAsync(IMigrationContext context, string migrationName)
+        {
+            // this is the de facto entry point via the migration runner, so create the table if it doesn't exist
```

Comment:
This can be done in multiple ways:
1) You make a private/protected method along the lines of `EnsureDbHasBeenInitialized`. You then call this method in all other methods (with this approach, it is best that it returns something that you need in those other methods, so it becomes "impossible" to forget - could be the db connection).
2) We add an `InitAsync` to the interface that the MigrationRunner then calls.

Reply:
fixed.